metadata-ingestion/src/datahub/ingestion/source/aws/dynamodb.py

```diff
@@ -0,0 +1,188 @@
+
+    @property
+    def dynamodb_client(self):
+        return self.get_dynamodb_client()
+
+    @property
+    def dynamodbstreams_client(self):
+        return self.get_dynamodbstreams_client()
+
+
+@dataclass
+class DynamoDBSourceReport(SourceReport):
+    filtered_tables: List[str] = field(default_factory=list)
+    streaming_disabled_tables: List[str] = field(default_factory=list)
+
+    def report_table_filtered(self, table: str) -> None:
+        self.filtered_tables.append(table)
+
+    def report_table_streaming_disabled(self, table: str) -> None:
+        self.streaming_disabled_tables.append(table)
+
+@dataclass
+class DynamoDBSource(Source):
+    config: DynamoDBSourceConfig
+    report: DynamoDBSourceReport()
+
+    def __init__(self, config: DynamoDBSourceConfig, ctx: PipelineContext):
+        super().__init__(ctx)
+        self.config = config
+        self.report = DynamoDBSourceReport()
+
+
+    @classmethod
+    def create(cls, source_config, context):
+        config = DynamoDBSourceConfig.parse_obj(source_config)
+        return cls(config, context)
+
+
+    def get_workunits(self) -> Iterable[MetadataWorkUnit]:
+        for table_name in get_all_tables(self.config.dynamodb_client):
+            logger.info(f"Processing table {table_name}")
+            if not self.config.table_pattern.allowed(table_name):
+                logger.warn(f"Table {table_name} is not allowed by table_allow pattern")
+                self.report.report_table_filtered(table_name)
+                continue
+            table = describe_table(self.config.dynamodb_client, table_name)
+            key_schema = table.get("KeySchema", [])
+            keys = [x["AttributeName"] for x in key_schema]
+            raw_attribute_definitions = table["AttributeDefinitions"]
+            attribute_definitions = {
+                x["AttributeName"]: x["AttributeType"] for x in raw_attribute_definitions
+            }
+            if table.get("StreamSpecification", {}).get("StreamEnabled", False):
+                stream_arn = table["LatestStreamArn"]
+                attribute_definitions.update(
+                    get_attribute_definitions(self.config.dynamodbstreams_client, stream_arn)
+                )
+            else:
+                logger.error(f"Table {table_name} streaming is NOT enabled.")
+                self.report.report_table_streaming_disabled(table_name)
+            dataset_urn = make_dataset_urn(DEFAULT_PLATFORM, table_name, self.config.env)
+            schema_fields = self.get_schema_fields(attribute_definitions)
+            schema_metadata = SchemaMetadata(
+                schemaName=f"{table_name}-schema",
+                platform=make_data_platform_urn(DEFAULT_PLATFORM),
+                version=0,
+                hash="",
+                platformSchema=KeyValueSchema(keySchema="", valueSchema=""),
+                fields=schema_fields,
+                primaryKeys=keys
```

Comment:
Is there a way to distinguish between partition and sort keys and not just put them both in primaryKeys?

Reply:
yeah I tried but didn't find a way. the `primaryKeys` is basically a list of string, so it does not allow us to annotate each key inside it.

Reply:
That is fine then. I was expecting to see something like `secondary key` which would be generic enough to apply to multiple use-cases.